go/internal/tinder/service.go

```diff
@@ -3,152 +3,141 @@ package tinder
-		o.FindPeerResetInterval = DefaultFindPeerInterval
-	}
+	return nil
 }
 
-func NewService(opts *Opts, h host.Host, drivers ...*Driver) (Service, error) {
-	// validate opts
-	opts.applyDefault()
+func (s *Service) WatchPeers(ctx context.Context, topic string) <-chan peer.AddrInfo {
+	return s.fadeOut(ctx, topic, 16)
+}
 
-	for _, driver := range drivers {
-		if err := driver.applyDefault(); err != nil {
-			return nil, fmt.Errorf("invalid driver %w", err)
+func (s *Service) fadeIn(ctx context.Context, topic string, in <-chan peer.AddrInfo) {
+	s.incProcess()
+	defer s.decProcess()
+
+	for {
+		select {
+		case p, ok := <-in:
+			if !ok {
+				return
+			}
+
+			if updated := s.peersCache.UpdatePeer(topic, p); updated {
+				s.logger.Debug("topic updated", zap.String("topic", topic), zap.String("peer", p.String()))
+			}
+		case <-ctx.Done():
+			return
 		}
 	}
+}
 
-	notify, err := NewNetworkUpdate(opts.Logger, h)
-	if err != nil {
-		return nil, err
-	}
-
-	// create monitor update
-	emitter, err := h.EventBus().Emitter(new(EvtDriverMonitor))
-	if err != nil {
-		return nil, err
-	}
+func (s *Service) fadeOut(ctx context.Context, topic string, bufsize int) <-chan peer.AddrInfo {
+	out := make(chan peer.AddrInfo, bufsize)
 
-	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		knownPeers := make(PeersUpdate)
 
-	wa := newWatchdogsAdvertiser(ctx, opts.Logger, h, notify, opts.AdvertiseResetInterval, opts.AdvertiseGracePeriod, drivers)
-	wd, err := newWatchdogsDiscoverer(ctx, opts.Logger, h, opts.FindPeerResetInterval, opts.BackoffStrategy, drivers)
-	if err != nil {
-		cancel()
-		return nil, err
-	}
+		for {
+			// Wait until `PeersCache` differ from `peerCache` peers status
+			updated, ok := s.peersCache.WaitForPeerUpdate(ctx, topic, knownPeers)
+			if !ok {
+				break
+			}
 
-	// compose discovery
-	var disc p2p_discovery.Discovery = &DriverDiscovery{
-		Advertiser: wa,
-		Discoverer: wd,
-	}
+			s.logger.Debug("got update on peer", zap.String("topic", topic), zap.Int("peers", len(updated)))
+			for _, peer := range s.peersCache.GetPeers(updated...) {
+				out <- peer
+			}
+		}
 
-	// enable monitor if needed
-	return &service{
-		Discovery:    disc,
-		Unregisterer: wa,
+		// we're done here, close the channel and decrement process
+		close(out)
+	}()
 
-		logger:  opts.Logger,
-		nnotify: notify,
-		emitter: emitter,
-		doneCtx: cancel,
-	}, nil
+	return out
 }
 
-func (s *service) Close() error {
-	s.doneCtx()
-
-	s.onceClose.Do(func() {
-		// we only want to log errors here, discard them on return
-		if err := s.emitter.Close(); err != nil {
-			s.logger.Warn("unable to close service emitter", zap.Error(err))
-		}
-		if err := s.nnotify.Close(); err != nil {
-			s.logger.Warn("unable to close network notify", zap.Error(err))
-		}
-	})
-
-	return nil
+func (s *Service) Close() error {
+	return s.networkNotify.Close()
 }
+
+func (s *Service) GetProcess() uint32 { return atomic.LoadUint32(&s.process) }
```

Comment:
I don't find where this public method is used.

Reply:
it should be used in a near future with prometheus metric